data_structures/dsu_path_compression.cpp

```diff
@@ -0,0 +1,183 @@
+         * and make a single set having a common representative.
+         * @param i element of some set
+         * @param j element of some set
+         * @returns void
+        */
+        void UnionSet(int i,int j){
+            //check if both belongs to same set or not
+            if(isSame(i,j)){
+                return;
+            }
+
+            //we find the representative of the i and j
+            int x = findSet(i);
+            int y = findSet(j);
+
+            //always keeping the min as x
+            //shallow tree
+            if(depth[x]>depth[y]){
+                std::swap(x,y);
+            }
+            //making the shallower root's parent the deeper root
+            p[x] = y;
+            
+            //if same depth then increase one's depth
+            if(depth[x] == depth[y]){
+                depth[y]++;
+            }
+            //total size of the resultant set.  
+            setSize[y] += setSize[x];
+            //changing the maximum elements
+            maxElement[y] = std::max(maxElement[x],maxElement[y]);
+            minElement[y] = std::min(minElement[x],minElement[y]);
+        }
+        /**
+         * @brief A utility function which check whether i and j belongs to
+         * same set or not
+         * @param i element of some set
+         * @param j element of some set
+         * @returns `true` if element `i` and `j` ARE in the same set
+         * @returns `false` if element `i` and `j` are NOT in same set
+         */
+        bool isSame(int i,int j){
+            if(findSet(i) == findSet(j)){
+                return true;
+            }
+            return false;
+        }
+        /**
+         * @brief prints the minimum, maximum and size of the set to which i belongs to
+         * @param i element of some set
+         * @returns void
+         */
+        void get(int i){
+            cout << "min:" << get_min(i) << " max:" << get_max(i) << " size of set:" <<size(i) << endl; 
+        }
+        /**
+         * @brief A utility function that returns the size of the set to which i belongs to
+         * @param i element of some set
+         * @returns size of the set to which i belongs to
+         */
+        int size(int i){
+            return setSize[findSet(i)];
+        }
+        /**
+         * @brief A utility function that returns the max element of the set to which i belongs to
+         * @param i element of some set
+         * @returns maximum of the set to which i belongs to
+         */
+        int get_max(int i){
+            return maxElement[findSet(i)];
+        }
+        /**
+         * @brief A utility function that returns the min element of the set to which i belongs to
+         * @param i element of some set
+         * @returns minimum of the set to which i belongs to 
+         */
+        int get_min(int i){
+            return minElement[findSet(i)];
+        }
+
+};
```

Comment:
Add self-test implementations using `assert` (see the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program as an example).